File: src/utils/strings.ts
import {KOTLIN_KEYWORDS} from "./constants.js";

export function capitalize(string: string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

export function camelize(string: string) {
    return string
        .replace(/([-_][A-Za-z])/g, group =>
            group
                .toUpperCase()
                .replace("-", "")
                .replace("_", "")
        )
        // remove leading and trailing delimiters
        .replace(/[-_]/g, "");
}

export function isValidIdentifier(string: string) {
    return (
        /^[\w$]+$/.test(string)
        && !/^\d/.test(string)
    )
}

export function escapeIdentifier(string: string) {
    if (KOTLIN_KEYWORDS.has(string)) {
        return `\`${string}\``
    }

    if (/^\d/.test(string)) {
        return `\`${string}\``
    }

    if (/^_+$/.test(string)) {
        return `\`${string}\``
    }

    return string
}

export function notEscapedIdentifier(string: string) {
    return camelize(
        string.replace(/\W/g, "-")
    )
}

export function identifier(string: string) {
    return escapeIdentifier(
        notEscapedIdentifier(string)
    )
}

export function constIdentifier(string: string) {
    return escapeIdentifier(
        string
            .replace(/\W/g, "_")
            .toUpperCase()
    )
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/util/strings.kt
package io.github.sgrishchenko.karakum.util

@JsExport
fun capitalize(string: String): String {
    if (string.isEmpty()) return string
    return string[0].uppercaseChar() + string.slice(1..string.lastIndex)
}

@JsExport
fun camelize(string: String): String {
    return string
        .replace("([-_][A-Za-z])".toRegex()) {
            it.value
                .uppercase()
                .replace("-", "")
                .replace("_", "")
        }
        // remove leading and trailing delimiters
        .replace("[-_]".toRegex(), "")
}

fun isKebab(string: String): Boolean {
    return "[\\w$-]+".toRegex().matches(string)
}

@JsExport
fun isValidIdentifier(string: String): Boolean {
    return "[\\w$]+".toRegex().matches(string)
            && !"^\\d".toRegex().containsMatchIn(string)
}

@JsExport
fun escapeIdentifier(string: String): String {
    if (string in KOTLIN_KEYWORDS) {
        return "`${string}`"
    }

    if ("_+".toRegex().matches(string)) {
        return "`${string}`"
    }

    if ("^\\d".toRegex().containsMatchIn(string)) {
        return "`${string}`"
    }

    return string
}

@JsExport
@Deprecated("Prefer escapeIdentifier")
fun identifier(string: String): String {
    return escapeIdentifier(
        camelize(
            string.replace("\\W".toRegex(), "-")
        )
    )
}

@JsExport
@Deprecated("Prefer escapeIdentifier")
fun constIdentifier(string: String): String {
    return escapeIdentifier(
        string
            .replace("\\W".toRegex(), "_")
            .uppercase()
    )
}
